### ISSUE TITLE
Is there any way to stop this numpy openblas/mkl dance?

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/numpy-feedstock#108]
Whenever I update anything I always have to specify numpy in the list as well, otherwise 

```
➤ conda install pytest
Solving environment: done

[...]

The following packages will be UPDATED:

    libgcc-ng:      7.2.0-hdf63c60_3                     conda-forge --> 8.2.0-hdf63c60_1     
    numpy:          1.15.2-py36_blas_openblashd3ea46f_1  conda-forge [blas_openblas] --> 1.15.2-py36h1d66e8a_1

The following packages will be DOWNGRADED:

    blas:           1.1-openblas                         conda-forge --> 1.0-mkl              
    opencv:         3.4.3-py36_blas_openblash829a850_200 conda-forge [blas_openblas] --> 3.4.1-py36h6fd60c2_1 
    scipy:          1.1.0-py36_blas_openblash7943236_201 conda-forge [blas_openblas] --> 1.1.0-py36hc49cb51_0 

```

```
conda install pytest numpy
```

doens't change the numpy version.

**Comment by msarahan** [comment]
add a constraint for blas=*=openblas

The problem is that conda really wants to get rid of the "feature" that is attached to that conda-forge package.  Implicit dependencies behave differently from explicit ones.  When you explicitly specify it, I believe the channel priority wins.  If you don't explicitly specify it, the feature minimization wins.

This may be helpful: https://conda.io/docs/user-guide/configuration/use-condarc.html#always-add-packages-by-default-create-default-packages

**Comment by hmaarrfk** [comment]
Thanks for the reply, 
Why does disallow not work? I tried adding `blas=*=mkl` and plain `mkl` there, and still tries to install `mkl`.

```
channels:
  - conda-forge
  - defaults

create_default_packages:
  - python=3.6
  - ipython

disallow:
  - mkl
```

**Comment by hmaarrfk** [comment]
For anybody else coming across this issue, I eventually ended up adding blas=*=openblas to `create_default_packages`. I figured I mostly require numpy anyway, so it isn't so bad.

The challenge comes if you already installed openblas (likely because of numpy) and now need to add a dependency for openblas. `conda install` will tell you it is already satisfied and not add it to the list of explicitely specified dependencies.

To fix that in your current environment
```
conda uninstall blas --force --offline
conda install blas=*=openblas --offline
```

`--force` because
* You just don't care about destroying your whole environment

`--offline` because
 * you are forcing uninstall, so who cares.
 * currently checking anaconda.org takes like 40 seconds and you likely already have openblas downloaded

maybe @msarahan has a better solution, but that one is pretty quick all things considered.

**Comment by ocefpaf** [comment]
@hmaarrfk can we close this one? I don't think there is anything we can do with the package to make that situation better. (But the workaround above seems to work just fine for me. Thanks for reporting that!)

**Comment by hmaarrfk** [comment]
If we close it, I think we would need to find a place to document the workaround. Having it as an issue makes the workaround easy to find.

Otherwise, we can make numpy depend on openblas?

**Comment by sbelharbi** [comment]
Thank you @hmaarrfk  for the workaround.

In my case: `conda install "blas=*=openblas" --offline` did not work (not sure why). I get:
 ```sh
$ conda install blas=*=openblas --offline
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - blas[build=openblas]

Current channels:

  - https://conda.anaconda.org/conda-forge/linux-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/free/linux-64
  - https://repo.anaconda.com/pkgs/free/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/pro/linux-64
  - https://repo.anaconda.com/pkgs/pro/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```

my `~/.condarc` contains:
```sh
channels:
  - conda-forge
  - defaults

create_default_packages:
  - python=3.7.1
  - blas=*=openblas

disallow:
  - mkl
```

In order to install `blas` with the build `openblas` i had to use ([#7548](https://github.com/conda/conda/issues/7548#issuecomment-405711038)) after removing `blas` with `conda uninstall blas --force --offline`:
```sh
conda install conda-forge::blas=*=openblas
```

**Comment by hmaarrfk** [comment]
@sbelharbi yeah, the `--offline` flag assumes that you had the package already in your cache.

I figured by explaining it you could have simply removed it.


In the docs: https://conda-forge.org/docs/blas.html#openblas-mkl-dance

I explicitly removed the `--offline` during the install command because I felt like it was a "micro-optimization" for the general user.

glad  you got it working.

**Comment by sbelharbi** [comment]
my bad!
I stopped reading the explanation at

> .. so who cares.

Thanks again!

**Comment by hmaarrfk** [comment]
Totally fine!

**Comment by mathause** [comment]
For me this problem was also resolved by setting

    channel_priority: strict

in my .condarc


(sorry for the double posting, I accidentally wrote it in the PR instead of here)

**Comment by russHyde** [comment]
Current url for the docs mentioned above: https://conda-forge.readthedocs.io/en/latest/blas.html#openblas-mkl-dance

**Comment by SteinsWesley** [comment]
In my situation, `conda uninstall blas --force  && conda install "blas=*=openblas"` didn't help, 
 `conda install nomkl numpy scipy scikit-learn numexpr` solves the problem, actually only nomkl installed.

ref:
https://stackoverflow.com/questions/51516740/conda-forcing-numpy-and-scipy-to-downgrade-from-openblas-to-non-working-mkl-vers

**Comment by hmaarrfk** [comment]
I think both instructions are defunct. Things changed alot with respect to blas recently.

Is conda updated for you? I haven't had the dance in a while

**Comment by SteinsWesley** [comment]
I'm in conda 4.6.14, Python 3.6.7 env,  I just installed it on a new machine, and encountered this problem,
when `conda update --all`
`The following packages will be UPDATED:

  blas                                         2.7-openblas --> 2.10-mkl
  libblas                                  3.8.0-7_openblas --> 3.8.0-10_mkl
  libcblas                                 3.8.0-7_openblas --> 3.8.0-10_mkl
  liblapack                                3.8.0-7_openblas --> 3.8.0-10_mkl
  liblapacke                               3.8.0-7_openblas --> 3.8.0-10_mkl
`
`when conda install sth.`
`The following packages will be DOWNGRADED:

  blas                                             2.10-mkl --> 2.7-openblas
  libblas                                      3.8.0-10_mkl --> 3.8.0-7_openblas
  libcblas                                     3.8.0-10_mkl --> 3.8.0-7_openblas
  liblapack                                    3.8.0-10_mkl --> 3.8.0-7_openblas
  liblapacke                                   3.8.0-10_mkl --> 3.8.0-7_openblas`
 `conda install nomkl numpy scipy scikit-learn numexpr` still works for me on this problem.

**Comment by jakirkham** [comment]
Chances are those packages have not been rebuilt with the new `blas` scheme. In particular I'm not sure if `numexpr` has. Would raise issues on the respective feedstocks.

Independently the scheme does not support `nomkl` AFAIAA. So if that is something you care about, would raise a separate issue for that.

In either case, it would be great for people to open new issues instead of commenting in an old closed one. These comments are likely to go unnoticed. New issues should make sure they get more attention. Thanks!
